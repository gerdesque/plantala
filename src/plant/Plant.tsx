import React from 'react';
import { Typography } from '@material-ui/core';

export interface IPlant {
  active: boolean
  name: string
  source: string
  amount?: number
  description?: string
  order ?: number
  rotation?: number
  scale?: number
  size?: number
}

export default function Plant({name, source, description}: IPlant) {

  return (
    <div className="plant">
      <Typography variant="h2">
        {name}
      </Typography>
    </div>
  )
}
