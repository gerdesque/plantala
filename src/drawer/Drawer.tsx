import React, {useState, useEffect, useContext} from 'react';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import { Avatar, Grid, Box, Slider, SwipeableDrawer, Typography } from '@material-ui/core';
import { IPlant } from '../plant/Plant';
import { AppContext } from '../app-plantala/Context';

interface IDrawerProps {
  activePlant: IPlant,
  setActivePlant: any,
  transformPlant: any
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    avatar: {
      width: theme.spacing(8),
      height: theme.spacing(8),
      borderColor: theme.palette.primary.main,
      borderStyle: 'solid',
      borderWidth: '2px',
      cursor: 'pointer'
    },
    drawer: {
      padding: theme.spacing(4),
      width: 200,
    }
  }),
);

export default function Drawer({activePlant, setActivePlant, transformPlant}: IDrawerProps) {
  const classes = useStyles();
  const [colorMode] = useContext(AppContext);
  const colorPath = colorMode ? '_Bunt' : '_SW';
  const [isOpen, setOpen] = useState(false);

  useEffect(() => {
    setOpen(activePlant.active || false);
  }, [activePlant]);

  const toggleDrawer = (open: boolean) => (
    event: React.KeyboardEvent | React.MouseEvent,
  ) => {
    if (
      event &&
      event.type === 'keydown' &&
      ((event as React.KeyboardEvent).key === 'Tab' ||
        (event as React.KeyboardEvent).key === 'Shift')
    ) {
      return;
    }
    setOpen(open);
    setActivePlant(null);
  };

  const selectedPlant = (
    <Grid container justify="center" className={classes.drawer} role="presentation">
      <Box>
      <Avatar
        key={activePlant.name}
        alt={activePlant.name}
        src={require('../assets/' + activePlant.source + colorPath + '.png').default}
        className={classes.avatar}
      />
      <Typography id="rotation-slider" gutterBottom>
        Rotation
      </Typography>
      <Slider
        aria-labelledby="rotation-slider"
        valueLabelDisplay="auto"
        max={360}
        value={activePlant.rotation || 0}
        onChange={(_event, newValue) => { transformPlant(activePlant, 'rotation', newValue) }}
      />
      <Typography id="rotation-slider" gutterBottom>
        Menge
      </Typography>
      <Slider
        aria-labelledby="amount-slider"
        valueLabelDisplay="auto"
        min={4}
        max={20}
        value={activePlant.amount || 8}
        onChange={(_event, newValue) => { transformPlant(activePlant, 'amount', newValue) }}
      />
      <Typography id="rotation-slider" gutterBottom>
        Größe
      </Typography>
      <Slider
        aria-labelledby="scale-slider"
        valueLabelDisplay="auto"
        min={0.1}
        step={0.1}
        max={2}
        value={activePlant.scale || 1}
        onChange={(_event, newValue) => { transformPlant(activePlant, 'scale', newValue) }}
      />
      <Typography id="distance-slider" gutterBottom>
        Abstand
      </Typography>
      <Slider
        aria-labelledby="distance-slider"
        valueLabelDisplay="auto"
        min={50}
        max={300}
        value={activePlant.distance || 50}
        onChange={(_event, newValue) => { transformPlant(activePlant, 'distance', newValue) }}
      />
      </Box>
    </Grid>
    );

  return (
    <SwipeableDrawer
      anchor="right"
      open={isOpen}
      onClose={toggleDrawer(false)}
      onOpen={toggleDrawer(true)}
    >
      {selectedPlant}
    </SwipeableDrawer>
  );
}
