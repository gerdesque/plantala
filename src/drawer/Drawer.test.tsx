import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import { Avatar, SwipeableDrawer, Slider } from '@material-ui/core';
import Drawer from './Drawer';
import { IPlant } from '../plant/Plant';

describe('Drawer', () => {

  let shallow: any;
  let wrapper:  any;
  const setActivePlant = jest.fn();
  const transformPlant = jest.fn();
  const activePlant = {name: 'Kardamom', source: 'A.png', selected: true, active: true}

  beforeEach(() => {
    shallow = createShallow();
    wrapper = shallow(<Drawer
      activePlant={activePlant}
      setActivePlant={setActivePlant}
      transformPlant={transformPlant}
      />);
  });

  it('should render correctly', () => expect(wrapper).toMatchSnapshot());

  it('should render a <SwipeableDrawer /> container', () => {
    expect(wrapper.find(SwipeableDrawer).length).toEqual(1);
  });

  it('should activate plant when <SwipeableDrawer /> component is opened', () => {
    wrapper.find(SwipeableDrawer).simulate('open');
    expect(setActivePlant).toHaveBeenCalledTimes(1);
  });
 
  it('should deactivate plant when <SwipeableDrawer /> component is closed', () => {
    wrapper.find(SwipeableDrawer).simulate('close');
    expect(setActivePlant).toHaveBeenCalledTimes(1);
  });

  it('should display the active <Avatar />', () => {
    expect(wrapper.find(Avatar)).toHaveLength(1);
  });

  it('should display 3 <Slider /> components', () => {
    expect(wrapper.find(Slider)).toHaveLength(3);
  });

  it('should transform active plant when rotation <Slider /> component has changed', () => {
    wrapper.find(Slider).at(0).simulate('change', undefined, 360);
    expect(transformPlant).toHaveBeenCalledWith(activePlant, 'rotation', 360);
  });

  it('should transform active plant when amount <Slider /> component has changed', () => {
    wrapper.find(Slider).at(1).simulate('change', undefined, 20);
    expect(transformPlant).toHaveBeenCalledWith(activePlant, 'amount', 20);
  });

  it('should transform active plant when scale <Slider /> component has changed', () => {
    wrapper.find(Slider).at(2).simulate('change', undefined, 2);
    expect(transformPlant).toHaveBeenCalledWith(activePlant, 'scale', 2);
  });

});
