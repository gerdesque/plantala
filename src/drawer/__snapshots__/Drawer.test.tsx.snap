// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer should render correctly 1`] = `
<ForwardRef(SwipeableDrawer)
  anchor="right"
  onClose={[Function]}
  onOpen={[Function]}
  open={false}
>
  <WithStyles(ForwardRef(Grid))
    className="makeStyles-drawer-2"
    container={true}
    justify="center"
    role="presentation"
  >
    <Styled(MuiBox)>
      <WithStyles(ForwardRef(Avatar))
        alt="Kardamom"
        className="makeStyles-avatar-1"
        key="Kardamom"
        src="/assets/A.png"
      />
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        id="rotation-slider"
      >
        Rotation
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Slider))
        aria-labelledby="rotation-slider"
        max={360}
        onChange={[Function]}
        value={0}
        valueLabelDisplay="auto"
      />
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        id="rotation-slider"
      >
        Menge
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Slider))
        aria-labelledby="amount-slider"
        max={20}
        min={4}
        onChange={[Function]}
        value={8}
        valueLabelDisplay="auto"
      />
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        id="rotation-slider"
      >
        Größe
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Slider))
        aria-labelledby="scale-slider"
        max={2}
        min={0.1}
        onChange={[Function]}
        step={0.1}
        value={1}
        valueLabelDisplay="auto"
      />
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        id="distance-slider"
      >
        Abstand
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Slider))
        aria-labelledby="distance-slider"
        max={300}
        min={50}
        onChange={[Function]}
        value={50}
        valueLabelDisplay="auto"
      />
    </Styled(MuiBox)>
  </WithStyles(ForwardRef(Grid))>
</ForwardRef(SwipeableDrawer)>
`;
